File: packages/text_reader/android/src/main/kotlin/com/jeffpoze/owned/textreader/TextReaderPlugin.kt
package com.jeffpoze.owned.textreader

import android.content.Context
import android.net.Uri
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.text.TextRecognition
import com.google.mlkit.vision.text.latin.TextRecognizerOptions
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.io.File

/** Reads printed text from an image file with Google ML Kit, on the device. */
class TextReaderPlugin : FlutterPlugin, MethodChannel.MethodCallHandler {
    private lateinit var channel: MethodChannel
    private lateinit var context: Context

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        channel = MethodChannel(binding.binaryMessenger, "owned/text_reader")
        channel.setMethodCallHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        if (call.method != "recognize") return result.notImplemented()
        val path = call.argument<String>("path") ?: return result.error("bad_image", "No image path.", null)
        val image = try {
            InputImage.fromFilePath(context, Uri.fromFile(File(path)))
        } catch (e: Exception) {
            return result.error("bad_image", "Couldn't open the image.", null)
        }
        val recognizer = TextRecognition.getClient(TextRecognizerOptions.DEFAULT_OPTIONS)
        recognizer.process(image)
            .addOnSuccessListener { text ->
                // Join pieces on the same printed row (a receipt's item and its price), top to bottom.
                val pieces = text.textBlocks.flatMap { it.lines }.filter { it.boundingBox != null }
                    .sortedBy { it.boundingBox!!.centerY() }
                val rows = mutableListOf<MutableList<com.google.mlkit.vision.text.Text.Line>>()
                for (p in pieces) {
                    val last = rows.lastOrNull()?.last()
                    val box = p.boundingBox!!
                    if (last != null && Math.abs(last.boundingBox!!.centerY() - box.centerY()) <
                        Math.min(last.boundingBox!!.height(), box.height()) / 2) {
                        rows.last().add(p)
                    } else {
                        rows.add(mutableListOf(p))
                    }
                }
                val lines = rows.map { row -> row.sortedBy { it.boundingBox!!.left }.joinToString("  ") { it.text } }
                result.success(lines.joinToString("\n"))
            }
            .addOnFailureListener { e -> result.error("recognition_failed", e.message, null) }
            .addOnCompleteListener { recognizer.close() }
    }
}

File: packages/text_reader/ios/text_reader/Sources/text_reader/TextReaderPlugin.swift
import Flutter
import UIKit
import Vision

/// Reads printed text from an image file with Apple's Vision framework, on the device.
public class TextReaderPlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "owned/text_reader", binaryMessenger: registrar.messenger())
    registrar.addMethodCallDelegate(TextReaderPlugin(), channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard call.method == "recognize",
          let args = call.arguments as? [String: Any],
          let path = args["path"] as? String else {
      result(FlutterMethodNotImplemented)
      return
    }
    guard let image = UIImage(contentsOfFile: path), let cgImage = image.cgImage else {
      result(FlutterError(code: "bad_image", message: "Couldn't open the image.", details: nil))
      return
    }

    let request = VNRecognizeTextRequest { request, error in
      if let error = error {
        DispatchQueue.main.async {
          result(FlutterError(code: "recognition_failed", message: error.localizedDescription, details: nil))
        }
        return
      }
      let observations = (request.results as? [VNRecognizedTextObservation]) ?? []
      let lines = TextReaderPlugin.rows(observations)
      DispatchQueue.main.async { result(lines.joined(separator: "\n")) }
    }
    request.recognitionLevel = .accurate
    // Model and serial numbers aren't words: don't "correct" them.
    request.usesLanguageCorrection = false

    let orientation = CGImagePropertyOrientation(image.imageOrientation)
    DispatchQueue.global(qos: .userInitiated).async {
      do {
        try VNImageRequestHandler(cgImage: cgImage, orientation: orientation, options: [:]).perform([request])
      } catch {
        DispatchQueue.main.async {
          result(FlutterError(code: "recognition_failed", message: error.localizedDescription, details: nil))
        }
      }
    }
  }
}

extension TextReaderPlugin {
  /// Joins pieces of text that sit on the same printed row (a receipt's item on the left
  /// and its price on the right), then orders rows top to bottom.
  static func rows(_ observations: [VNRecognizedTextObservation]) -> [String] {
    // Vision's y axis points up, so higher midY means nearer the top.
    let sorted = observations.sorted { $0.boundingBox.midY > $1.boundingBox.midY }
    var rows: [[VNRecognizedTextObservation]] = []
    for o in sorted {
      if let last = rows.last?.last,
         abs(last.boundingBox.midY - o.boundingBox.midY) < min(last.boundingBox.height, o.boundingBox.height) * 0.5 {
        rows[rows.count - 1].append(o)
      } else {
        rows.append([o])
      }
    }
    return rows.map { row in
      row.sorted { $0.boundingBox.minX < $1.boundingBox.minX }
        .compactMap { $0.topCandidates(1).first?.string }
        .joined(separator: "  ")
    }
  }
}

extension CGImagePropertyOrientation {
  init(_ o: UIImage.Orientation) {
    switch o {
    case .up: self = .up
    case .down: self = .down
    case .left: self = .left
    case .right: self = .right
    case .upMirrored: self = .upMirrored
    case .downMirrored: self = .downMirrored
    case .leftMirrored: self = .leftMirrored
    case .rightMirrored: self = .rightMirrored
    @unknown default: self = .up
    }
  }
}
